lte/gateway/c/session_manager/LocalSessionManagerHandler.cpp

```diff
@@ -250,97 +250,137 @@ void LocalSessionManagerHandlerImpl::CreateSession(
-    switch (common.rat_type()) {
-      case TGPP_WLAN:
-        action = handle_create_session_cwf(session_map, session_id, cfg);
-        break;
-      case TGPP_LTE:
-        action = handle_create_session_lte(session_map, session_id, cfg);
-        break;
-      default:
-        // this should be handled above
-        MLOG(MERROR) << "RAT type " << common.rat_type() << " is not supported";
-        return;
-    }
-    if (action.end_existing_session) {
-      // Assumption here is that sessions are unique by IMSI+APN
-      end_session(
-          session_map, common.sid(), common.apn(),
-          [&](grpc::Status status, LocalEndSessionResponse response) {});
-    }
-    if (action.create_new_session) {
-      send_create_session(
-          session_map, action.session_id_to_send_back, cfg, response_callback);
-    }
-    if (action.status_back_to_access) {
-      send_local_create_session_response(
-          *(action.status_back_to_access), action.session_id_to_send_back,
-          response_callback);
-    }
-  });
+        log_create_session(cfg);
+        auto status = validate_create_session_request(cfg);
+        if (!status.ok()) {
+          send_local_create_session_response(status, "", response_callback);
+          return;
+        }
+
+        const auto& session_id = id_gen_.gen_session_id(imsi);
+        auto session_map       = session_store_.read_sessions({imsi});
+        SessionActionOrStatus action;
+        switch (common.rat_type()) {
+          case TGPP_WLAN:
+            action = handle_create_session_cwf(session_map, session_id, cfg);
+            break;
+          case TGPP_LTE:
+            action = handle_create_session_lte(session_map, session_id, cfg);
+            break;
+          default:
+            // this should be handled above
+            MLOG(MERROR) << "RAT type " << common.rat_type()
+                         << " is not supported";
+            return;
+        }
+        if (action.end_existing_session) {
+          // Assumption here is that sessions are unique by IMSI+APN
+          end_session(
+              session_map, common.sid(), common.apn(),
+              [&](grpc::Status status, LocalEndSessionResponse response) {});
+        }
+        if (action.create_new_session) {
+          bool success = initialize_session(
+              session_map, action.session_id_to_send_back, cfg);
+          if (success) {
+            send_create_session(
+                session_map, action.session_id_to_send_back, cfg,
+                response_callback);
+          } else {
+            // abort and send back a failure response to access
+            action.set_status(
+                Status(grpc::ABORTED, "Failed to update SessionStore"));
+          }
+        }
+        if (action.status_back_to_access) {
+          send_local_create_session_response(
+              *(action.status_back_to_access), action.session_id_to_send_back,
+              response_callback);
+        }
```

Comment:
I may be missing the point, but I don't understand the logic here. Is action being used for multiple things? 

Action was the object that defined if a session was in a creating state. But here it seems that it also defines if we have to provide a response to MME? Wouldn't it be better just use booleans and different variables.

Reply:
Maybe the naming is confusing here. It was more of a way to wrap multiple return values into one struct. 
Some things to consider are:
1. Depending on whether we're creating a new session or recycling, the session ID we need to return is different
2. Sometimes we need to end existing sessions for the IMSI+APN pair before sending a request
3. Sometimes we just need to return to MME directly either due to there already being a session (recycling) or some other failure

I wanted a way to return all those information in one struct since whether we decide to recycle or not depends on different logic for each rat type.

Reply:
that's why I called it SessionActionOrStatus so that it contains either 
1. grpc::Status that needs to be returned to MME right away 
2. different actions that need to be taken : end session / create session in policydb , etc.

Reply:
I'll try this clean this up a bit though.

Reply:
Sure. 

I still think we should be better capturing the state using `ACTIVATING` flag.

But as long as it works i am fine

Reply:
Yeah let me attempt using the session state flag. Might be a cleaner approach.